channels/api.py

```diff
@@ -56,26 +69,97 @@ def get_or_create_user(username):
     refresh_token_url = urljoin(settings.OPEN_DISCUSSIONS_REDDIT_URL, '/api/v1/generate_refresh_token')
 
     session = _get_session()
-    resp = session.get(refresh_token_url, params={'username': username}).json()
-    return resp['refresh_token']
+    return session.get(refresh_token_url, params={'username': username}).json()
+
 
+def get_auth_tokens(user):
+    """
+    Gets the stored refresh token or generates a new one
 
-def _get_user_credentials(user):
+    Args:
+        user (User): the authenticated user
+
+    Returns:
+        (channels.models.RedditRefreshToken, channels.models.RedditAccessToken): the stored tokens
     """
-    Get credentials for authenticated user
+    threshold_date = datetime.now(pytz.utc) + timedelta(minutes=2)
+    refresh_token, _ = RedditRefreshToken.objects.get_or_create(user=user)
+    access_token = None
+
+    # if we created this token just now, atomically generate one
+    if not refresh_token.token_value:
```

Comment:
The record could get created, but if the reddit backend fails then you'll get stuck with a refresh token with no value. I really wanted to avoid a transaction here too.